Fix for issue 34185 ovs fail-mode

ashish-billore opened the pull request:
This patch fixes issue #34185
This is related to ovs fail-mode when there is already a fail-mode on the ovs and appears in 'have' and unspecified or none in the request ('want'), as per the default in ovs.

Signed-off-by: Ashish Billore <ashish.billore@gmail.com>

##### SUMMARY
<!--- Describe the change below, including rationale and design decisions -->
This is related to ovs fail-mode when there is already a fail-mode on the ovs and appears in 'have' and unspecified or none in the request ('want'), as per the default in ovs.

https://github.com/ansible/ansible/issues/34185
<!--- HINT: Include "Fixes #nnn" if you are fixing an existing issue -->
Fixes #34185
##### ISSUE TYPE
<!--- Pick one below and delete the rest -->
- Bugfix Pull Request


##### COMPONENT NAME
<!--- Write the short name of the module, plugin, task or feature below -->
openvswitch_bridge_module
https://docs.ansible.com/ansible/2.6/modules/openvswitch_bridge_module.html

##### ANSIBLE VERSION
<!--- Paste verbatim output from "ansible --version" between quotes -->
```paste below
ansible 2.4.0.0 and above
```

##### ADDITIONAL INFORMATION
<!--- Include additional information to help people understand the change here -->

This patch adds a check to see if there is already fail-mode set on the ovs and the request specified has the fail-mode or not. If the fail-mode is specified in the request it works as earlier, just apply the fail-mode in 'want'. If there is no fail-mode specified in the 'want' or the request and if there is already a fail-mode set on the ovs, it resets the fail-mode on the ovs bridge, which is the default in ovs.

<!--- A step-by-step reproduction of the problem is helpful if there is no related issue -->

Please refer: #34185

<!--- Paste verbatim command output below, e.g. before and after your change -->
```paste below

Before the patch is applied:

[root@osc]# ansible master -i inventory.yaml -m openvswitch_bridge -a "bridge=br-int state=present"
128.90.0.54 | FAILED! => {
    "changed": false, 
    "cmd": "/usr/bin/ovs-vsctl -t 5 set-fail-mode br-int None", 
    "failed": true, 
    "msg": "ovs-vsctl: fail-mode must be \"standalone\" or \"secure\"", 
    "rc": 1, 
    "stderr": "ovs-vsctl: fail-mode must be \"standalone\" or \"secure\"\n", 
    "stderr_lines": [
        "ovs-vsctl: fail-mode must be \"standalone\" or \"secure\""
    ], 
    "stdout": "", 
    "stdout_lines": []
}

After applying the path:

[root@osc]# ansible master -i inventory.yaml -m openvswitch_bridge -a "bridge=br-int state=present"
128.90.0.54 | SUCCESS => {
    "changed": true, 
    "commands": [
        "/usr/bin/ovs-vsctl -t 2 del-fail-mode br-int"
    ], 
    "failed": false
}
[root@osc]#
```

ashish-billore committed:
Fix for issue 34185 ovs fail-mode

    This patch fixes issue# 34185
    related to ovs fail-mode when there is already a fail-mode
    in 'have' and unspecified or none in 'want', as per the
    default in ovs.

Signed-off-by: Ashish Billore <ashish.billore@gmail.com>

ashish-billore commented:
Does this fix look reasonable to be merged or there are any suggestions?

no2a commented:
@ashish-billore: "No-op" in the documentation string of fail_mode means "leave it untouched", IMO. So when fail_mode is not specified, neither set-fail-mode nor del-fail-mode should be performed.

ashish-billore commented:
@no2a: Thanks for the suggestion.

I can think of following cases that need to be handled in existing logic:

Already fail_mode is set ('have') on the bridge and the request ('want') has:
1. Unspecified (None): In this case we can go with "No-op" with just retaining the existing fail_mode.
2. Use wants to clear the existing 'fail_mode' using 'del-fail-mode'. In this case I can think of 2 ways to implement it:
- Add another value to indicate this:
Something like--> choices : [secure, standalone, clear_fail_mode]
- Add another 'options':
Something like: 'del_fail_mode' which will allow user to clear off existing fail_mode.

Please give your suggestions.
Thanks!

no2a commented on lib/ansible/modules/network/ovs/openvswitch_bridge.py:
```diff
@@ -123,12 +123,23 @@ def map_obj_to_commands(want, have, module):
+                                               " set-fail-mode %(bridge)s"
+                                               " %(fail_mode)s")
+                    else:
+                        # if fail-mode is specified, but empty it means user
```
Clearing fail-mode is a new feature. Could you update description of the fail_mode in DOCUMENTATION?

no2a reviewed (changes requested):
The code is LGTM. DOCUMENTATION also should be updated though.

ashish-billore commented:
@no2a Thanks for checking this patch.
I'll perhaps separate the fix for this issue and the new feature of clearing fail-mode in 2 patches and submit it with doc update as well. This will keep the patch consistent with the issue headline.

I'll submit these 2 patches shortly, thanks!

ashish-billore commented:
@no2a I have reverted the new feature of clearing off the fail_mode in this new patch. Please see if this is fine for fixing this issue.
I'll submit another separate patch for Clearing off the fail_mode feature with updated documentation.
Thanks!

ashish-billore committed:
Add support for clearing off fail_mode

Implemet the feature of clearing off previously set
fail_mode by passing empty fail_mode.

ashish-billore committed:
Add support for clearing off fail_mode

Implemet the feature of clearing off previously set
fail_mode by passing empty fail_mode.

gdpak reviewed (changes requested):
fail_mode=None should be no_op for backward compatability

ashish-billore commented:
These patches now handle case where fail_mode is passed as 'None'. Please review. Thanks for the help.

ashish-billore commented:
@gdpak Can you please check if #56c9cf8d70e833f53f93f9983ba53ed9a10bbc82 looks fine now?

gdpak commented on lib/ansible/modules/network/ovs/openvswitch_bridge.py:
```diff
@@ -123,12 +125,23 @@ def map_obj_to_commands(want, have, module):
-                                       " %(fail_mode)s")
-                command = templatized_command % module.params
-                commands.append(command)
+            wanted_fail_mode = want['fail_mode']
```
@ashish-billore  It looks much better now. Thanks for making changes. One comment here is passing empty value to ansible module arg does not look good. Probably we can pass something as 'clear' to convey user intention to clear previously set fail_mode. Also fail_mode is a choice parameter in doc string so it has to be a choice parameter in code as well. This also need fixing. I could test following code to be working as expected. please take a look
```diff
--- a/lib/ansible/modules/network/ovs/openvswitch_bridge.py
+++ b/lib/ansible/modules/network/ovs/openvswitch_bridge.py
@@ -53,9 +53,11 @@ options:
               To  clear all external-ids pass an empty value.
     fail_mode:
         version_added: 2.0
-        choices : [secure, standalone]
+        choices : [secure, standalone, clear]
         description:
             - Set bridge fail-mode. The default value (None) is a No-op.
+              Omitting this parameter is a No-op.
+              To clear previously set fail_mode, pass it as 'clear'.
     set:
         version_added: 2.3
         description:
@@ -123,12 +125,23 @@ def map_obj_to_commands(want, have, module):
             commands.append(command)
     else:
         if have:
-            if want['fail_mode'] != have['fail_mode']:
-                templatized_command = ("%(ovs-vsctl)s -t %(timeout)s"
-                                       " set-fail-mode %(bridge)s"
-                                       " %(fail_mode)s")
-                command = templatized_command % module.params
-                commands.append(command)
+            wanted_fail_mode = want['fail_mode']
+            have_fail_mode = have['fail_mode']
+            if wanted_fail_mode != have_fail_mode:
+                if wanted_fail_mode is not None:
+                    if wanted_fail_mode != 'clear':
+                        # only set the fail-mode if specified and not empty
+                        templatized_command = ("%(ovs-vsctl)s -t %(timeout)s"
+                                               " set-fail-mode %(bridge)s"
+                                               " %(fail_mode)s")
+                    else:
+                        # if fail-mode is specified, but empty it means user
+                        # wants to reset the fail-mode, so clear it off
+                        templatized_command = ("%(ovs-vsctl)s -t %(timeout)s"
+                                               " del-fail-mode %(bridge)s")
+
+                    command = templatized_command % module.params
+                    commands.append(command)
 
             if want['external_ids'] != have['external_ids']:
                 templatized_command = ("%(ovs-vsctl)s -t %(timeout)s"
@@ -233,7 +246,8 @@ def main():
         'state': {'default': 'present', 'choices': ['present', 'absent']},
         'timeout': {'default': 5, 'type': 'int'},
         'external_ids': {'default': None, 'type': 'dict'},
-        'fail_mode': {'default': None},
+        'fail_mode': {'default': None, 'choices': ['secure', 'standalone',
+            'clear']},
         'set': {'required': False, 'default': None}
     }
 
```

ashish-billore replied:
thanks @gdpak looks good to me, minor updates to the comment in the above else check for clear:
```suggestion
                    else:
                       # if fail-mode is specified as 'clear' means user
                       # wants to reset the fail-mode, so clear it off
```

ashish-billore reviewed (commented):
Looks fine to me, minor updates in comments for the 'clear' off else check

ashish-billore commented:
@no2a @gdpak @gundalow thanks for your interest in this long open PR and providing valuable comments. If it looks fine to cover needed cases, please advise to take it forward.
Thanks!

webknjaz commented on lib/ansible/modules/network/ovs/openvswitch_bridge.py:
```diff
@@ -123,12 +125,23 @@ def map_obj_to_commands(want, have, module):
+            wanted_fail_mode = want['fail_mode']
+            have_fail_mode = have['fail_mode']
+            if wanted_fail_mode != have_fail_mode:
+                if wanted_fail_mode is not None and wanted_fail_mode != 'None':
```
Please merge this clause with previous if-clause using `and`, this will reduce the whole block indentation. Less nesting is better.

gdpak commented:
@ashish-billore  Can you please commit changes, which we agreed to this PR . Let's get it in devel.

abenokraitis commented:
@ashish-billore Any updates? Thanks! needs_info

ashish-billore commented:
@abenokraitis Sorry for the delay on this, will submit updated patch shortly.
Thanks!

Qalthos commented:
Hey, we can't proceed on this without the changes. If you could update the PR, we can move forward on this. Thanks.

needs_info

pabelanger commented:
recheck

pabelanger commented:
recheck

pabelanger reviewed (changes requested):
I'd like to see us update tests/integration/targets/openvswitch_bridge and add a new test to validate this code.

We now have pre-merge tests running from zuul.ansible.com, so help ensure it is properly working before we merge this.

ashish-billore commented:
recheck

ashish-billore commented:
recheck

ashish-billore commented:
recheck

danielmellado commented:
Thank you very much for your interest in Ansible. This plugin is no longer maintained in this repository and has been migrated to https://github.com/ansible-collections/openvswitch.openvswitch
Please re-submit this PR in the above repository.
If you have further questions please stop by IRC or the mailing list:

IRC: #ansible-network on irc.freenode.net
mailing list: https://groups.google.com/forum/#!forum/ansible-project
